File: java/src/edu/berkeley/cs/sort/Merge.java
package edu.berkeley.cs.sort;

import edu.berkeley.cs.util.LinkedList;
import java.util.Arrays;

public class Merge<T extends Comparable> extends Sort<T> {
  protected Insertion<T> insertion = new Insertion<>();

  @Override
  public void sort(T[] input) {
    T[] auxiliary = Arrays.copyOf(input, input.length);
    sort(input, auxiliary, 0, input.length - 1);
  }

  private void merge(T[] input, T[] auxiliary, int low, int mid, int high) {}

  private void sort(T[] input, T[] auxiliary, int low, int high) {}

  private LinkedList<T> merge(LinkedList<T> left, LinkedList<T> right) {
    return null;
  }

  public LinkedList<T> sort(LinkedList<T> input) {
    return null;
  }
}

File: python/src/sort/merge.py
from python.src.sort.insertion import Insertion
from python.src.sort.sort import Sort


class Merge(Sort):

    def __init__(self):
        super().__init__()
        self.insertion = Insertion()

    def merge(self, input, auxiliary, low, mid, high):
        pass

    def sort(self, input, low=None, high=None):
        auxiliary = list(input)

        if low is None:
            low = 0

        if high is None:
            high = len(input) - 1

        self.__sort(input, auxiliary, low, high)

    def __sort(self, input, auxiliary, low, high):
        pass

    def linkedlist_merge(self, left, right):
        pass

    def linkedlist_sort(self, input):
        return None
